File: cpp/cpp-boost/demo06b-list-threads.cpp
/*
LIST OF MULTIPLE THREADS
Version B: Using the std::vector
*/


#include <iostream>
#include <vector>
#include <boost/shared_ptr.hpp>
#include <boost/thread.hpp>
using namespace std;



typedef boost::shared_ptr<boost::thread> threadptr;



void doTask(int index) {
    cout << index;
}



int main() {
    const int NUM_THREADS = 5;

    vector<threadptr> lstTh;

    for (int i = 0; i < NUM_THREADS; ++i) {
        threadptr ptr = boost::make_shared<boost::thread>(doTask, i);
        lstTh.push_back(ptr);
    }

    for (int i = 0; i < lstTh.size(); ++i) {
        lstTh[i]->join();
    }

    cout << endl;
    return 0;
}

File: js-nodejs/demo06b-list-threads.js
/*
LIST OF MULTIPLE THREADS
VERSION B: Using mylib.createThread
*/

import * as mylib from './mylib.js';
import { isMainThread, workerData } from 'worker_threads';


//---------------------------------------------
//            WORKER THREAD SECTION
//---------------------------------------------


const doTask = index => {
  console.log(index);
};

const workerFunc = () => {
  try {
    doTask(workerData);
  } catch (error) {
    console.error(error);
  }
};


//---------------------------------------------
//             MAIN THREAD SECTION
//---------------------------------------------


const mainFunc = async () => {
  try {

    const NUM_THREADS = 5;
    const lstTh = new Array(NUM_THREADS);

    for (let i = 0; i < NUM_THREADS; ++i) {
      lstTh[i] = mylib.createThread(new URL(import.meta.url), i);
    }

    await Promise.all(lstTh.map(([_,pr]) => pr));
    console.log('Good bye');

  } catch (error) {
    console.error(error);
  }
};


if (isMainThread) {
  await mainFunc();
} else {
  workerFunc();
}
